refactor integration tests

jroebu14 opened the pull request:
**Overall change:**
Removes the need to pass in hard coded strings to identify things are in the document. Uses snapshots to create a baseline we can test against to catch any regressions. Removes a lot of the boilerplate set up in test `.test.js` file. Removes DOM testing library in favour of query selectors. Removed full page snapshots. Prepares integration tests to be run across all 41 services https://github.com/bbc/simorgh/pull/6228

**Code changes:**

Refactored pages directory structure and test file setup. The structure is now:

```
── pages
|  ├── articles
|  |  ├── ampTests.js
|  |  ├── canonicalTests.js
|  |  ├── crossPlatformTests.js
|  |  ├── mundo
|  |  |  ├── __snapshots__
|  |  |  ├── amp.test.js
|  |  |  └── canonical.test.js
|  |  └── pidgin
|  |     ├── __snapshots__
|  |     ├── amp.test.js
|  |     └── canonical.test.js
|  ├── frontPage
|  |  ├── ampTests.js
|  |  ├── arabic
|  |  |  ├── __snapshots__
|  |  |  ├── amp.test.js
|  |  |  └── canonical.test.js
|  |  ├── canonicalTests.js
|  |  ├── crossPlatformTests.js
|  |  └── pidgin
|  |     ├── __snapshots__
|  |     ├── amp.test.js
|  |     └── canonical.test.js
|  ├── liveRadio
|  |  ├── amharic
|  |  |  ├── __snapshots__
|  |  |  ├── amp.test.js
|  |  |  └── canonical.test.js
|  |  ├── ampTests.js
|  |  ├── canonicalTests.js
|  |  ├── crossPlatformTests.js
|  |  └── korean
|  |     ├── __snapshots__
|  |     ├── amp.test.js
|  |     └── canonical.test.js
|  └── storyPage
|     ├── crossPlatformTests.js
|     └── mundo
|        ├── __snapshots__
|        ├── amp.test.js
|        └── canonical.test.js
```

More info on the new setup can be found in the [README](https://github.com/bbc/simorgh/blob/644507719ca7e43126a4bff7c60f67ee60c85983/src/integration/README.md).

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added labels to this PR for the relevant pod(s) affected by these changes
- [x] I have assigned this PR to the Simorgh project

**Testing:**

- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`)
- [ ] This PR requires manual testing

Commit:
refactor integration tests

karinathomasbbc commented on package.json:
```diff
@@ -24,16 +24,16 @@
     "test:unit:watch": "npm run test:unit -- --watch",
-    "test:integration:ci": "jest --colors ./src/integration/pages/**/*.test.js",
-    "test:integration": "./scripts/killApp.sh && npm run build && (npm run start & wait-on -t 20000 http://localhost:7080/status) && jest --colors ./src/integration/pages/**/*.test.js",
+    "test:integration": "./scripts/killApp.sh && npm run build && (npm run start & wait-on -t 20000 http://localhost:7080/status) && jest --colors ./src/integration/pages/**/*",
```
Can we add a couple of params for optimisation:
`jest --runInBand --colors ./src/integration/pages/**/*"`

jroebu14 replied:
Cool. What does it do?

karinathomasbbc replied:
https://jestjs.io/docs/en/cli.html#--runinband

Run all tests serially in the current process, rather than creating a worker pool of child processes that run tests.

All of the other jest commands seem to use this (see `test:unit`), and I recall a discussion on Slack somewhere too regarding performance maybe?

karinathomasbbc commented on package.json:
```diff
@@ -24,16 +24,16 @@
     "storybook": "start-storybook -p 9001 -s .storybook/static -c .storybook",
     "test": "npm run test:lint && npm run test:dependencies && npm run test:unit && npm run build -- --hide-modules --colors && run-p --race start amp:validate && run-p --race start test:integration:ci",
-    "test:ci": "npm run start & npm install --no-save cypress@3.4.1 bbc-a11y@latest npm-run-all@latest puppeteer@latest audit-ci@latest && npm run test:integration:ci && run-p cypress:ci test:puppeteer && run-p bbcA11y:ci lighthouse && audit-ci --low --whitelist acorn minimist",
+    "test:ci": "npm run start & npm install --no-save cypress@3.4.1 bbc-a11y@latest npm-run-all@latest puppeteer@latest audit-ci@latest && npm run test:integration:ci && npm run test:integration:all-services:ci && run-p cypress:ci test:puppeteer && run-p bbcA11y:ci lighthouse && audit-ci --low --whitelist acorn minimist",
```
`npm run test:integration:all-services:ci` doesn't exist on this branch - might need to remove for now.

karinathomasbbc commented on package.json:
```diff
@@ -24,16 +24,16 @@
-    "test:integration:ci": "jest --colors ./src/integration/pages/**/*.test.js",
-    "test:integration": "./scripts/killApp.sh && npm run build && (npm run start & wait-on -t 20000 http://localhost:7080/status) && jest --colors ./src/integration/pages/**/*.test.js",
+    "test:integration": "./scripts/killApp.sh && npm run build && (npm run start & wait-on -t 20000 http://localhost:7080/status) && jest --colors ./src/integration/pages/**/*",
+    "test:integration:ci": "jest --colors ./src/integration/pages/**/*",
```
And to here too:
` jest --ci --runInBand --colors ./src/integration/pages/**/*"`

karinathomasbbc replied:
https://jestjs.io/docs/en/cli.html#--ci

Jest will assume it is running in a CI environment. This changes the behavior when a new snapshot is encountered. Instead of the regular behavior of storing a new snapshot automatically, it will fail the test and require Jest to be run with `--updateSnapshot`

Commit:
Merge branch 'latest' into refactor-integration-tests

Commit:
update scrips in package.json

Commit:
Merge branch 'latest' into refactor-integration-tests

Commit:
refactor to one test example per page type

Commit:
Merge branch 'latest' into refactor-integration-tests

Commit:
refactor to add more test examples back in

Commit:
update snapshot

Commit:
add header tests

Commit:
use correct footer copyright test title

Commit:
separate snapshots for each platform

jroebu14 commented:
Sorry @karinathomasbbc I was still working on this branch which is why it was in draft. I made a branch with your changes https://github.com/bbc/simorgh/pull/6227

Commit:
update snapshots

Commit:
Merge branch 'latest' into refactor-integration-tests

Commit:
update snapshots

Commit:
update test titles

karinathomasbbc commented:
Oops! Meant to close https://github.com/bbc/simorgh/pull/6227 instead.

Commit:
make utils folder. add unit tests

Commit:
use module.exports

Commit:
update readme

Commit:
add some clarity on the AMP url

andrewscfc commented on src/integration/README.md:
```diff
@@ -14,38 +14,90 @@ This will build and run the application that the tests will run against.
-In the above example we have specified a pathname using a [docblock pragma](#what-is-a-docblock-pragma). The pathname is the part of the url that is everything after `https://bbc.com` and in this example it is `/mundo/articles/ce42wzqr2mko`. If you visit `https://bbc.com/mundo/articles/ce42wzqr2mko` you will see this is a Mundo article page and it is what we are going to test.
+#### AMP only tests
+
+For tests that should be run only on the AMP platform, we should add these tests to a file called `ampTests.js`. An example of a cross platform test would be like
```
```suggestion
For tests that should be run only on the AMP platform, we should add these tests to a file called `ampTests.js`. An example of an AMP test would be like
```

andrewscfc commented on src/integration/README.md:
```diff
@@ -14,38 +14,90 @@ This will build and run the application that the tests will run against.
+
+#### Canonical only tests
+
+For tests that should be run only on the canonical platform, we should add these tests to a file called `canonicalTests.js`. An example of a cross platform test would be like
```
```suggestion
For tests that should be run only on the canonical platform, we should add these tests to a file called `canonicalTests.js`. An example of a canonical test would be like
```

andrewscfc commented on src/integration/README.md:
```diff
@@ -14,38 +14,90 @@ This will build and run the application that the tests will run against.
+- `pageType` - The page type which is parsed from the test file path.
+- `service` - The service of the currently running page test.
+- `document` - The `document` object of the current page. You can use this to query the DOM and assert the correct things are in the page.
+- `platform` - The current platform (`amp` or `canonical`)
```
Is there any need to have this in the global scope? Isn't that implicit knowledge based on whether you are writing and AMP or canonical test? And I guess a cross platform test shouldn't matter. 

I'm just wary of cross-platform tests containing `if(platform === 'amp')` rather than tests being correctly put in their respective platform related files.

jroebu14 replied:
that's a fair point!

andrewscfc reviewed (commented):
Quick feedback on the readme, gonna go over the implementation next 👟

andrewscfc commented on jest.config.js:
```diff
@@ -1,65 +1,60 @@
-      transform: {
-        '^.+\\.js$': 'babel-jest',
-        '^.+\\.jsx$': 'babel-jest',
+  projects: [unitTests, ampIntegrationTests, canonicalIntegrationTests],
```
Does [this regex above](https://github.com/bbc/simorgh/pull/6168/files#diff-2d0cd5d10b9604941c38c6aac608178aR22) capture amp and canonical tests for the `unitTests` configuration?

If it does I think it means the amp and canonical platform tests will run twice, once under the `unitTests` config and once for the respective platform configs (`ampIntegrationTests`, `canonicalIntegrationTests`).

karinathomasbbc replied:
@andrewscfc We have already excluded integration tests from the unit test command (`test:unit`) in package.json https://github.com/bbc/simorgh/blob/ea216180db8ab8d2f17c8352879e78bae86c33f0/package.json#L34

andrewscfc replied:
Okay, probably not understanding. If we are excluding integration tests in the command, why are we including them in this config: https://github.com/bbc/simorgh/pull/6168/files/04fda4d821dc7b03ffde9135cc7cc0213c326674#diff-2d0cd5d10b9604941c38c6aac608178aR22 

I'm sure I'm missing something, apologies

jroebu14 replied:
That glob excludes the integration tests except for anything in the `utils` folder of integration tests because they contain unit tests. They don't test an integration, they literally test individual units of code.

Commit:
readme updates

Commit:
readme updates

Commit:
readme updates

Commit:
readme updates

Commit:
update snapshots

Commit:
Merge branch 'latest' into refactor-integration-tests

Commit:
fix lint issues

andrewscfc commented on package.json:
```diff
@@ -33,8 +33,8 @@
     "test:unit:watch": "npm run test:unit -- --watch",
-    "test:integration:ci": "jest --ci --runInBand --colors ./src/integration/pages/**/*.test.js",
-    "test:integration": "./scripts/killApp.sh && npm run build:local && (npm run start & wait-on -t 20000 http://localhost:7080/status) && jest --runInBand --colors ./src/integration/pages/**/*.test.js",
+    "test:integration": "./scripts/killApp.sh && npm run build:local && (npm run start & wait-on -t 20000 http://localhost:7080/status) && jest --no-cache --runInBand --colors ./src/integration/pages/**/*",
```
Perhaps not in scope for this PR but AFAIK we still have the issue where a zombie simorgh app is left running once the integration tests have finished, I've not tried this suggestion but I'm fairly confident it'll fix it.
```suggestion
    "test:integration": "./scripts/killApp.sh && npm run build:local && (npm run start & wait-on -t 20000 http://localhost:7080/status) && (jest --no-cache --runInBand --colors ./src/integration/pages/**/*) && ./scripts/killApp.sh",
```

jroebu14 replied:
@karinathomasbbc I've been meaning to chat to you about how to solve this

jroebu14 replied:
I'm happy to make a separate PR with these changes @andrewscfc. Thanks

andrewscfc commented on package.json:
```diff
@@ -33,8 +33,8 @@
-    "test:integration:ci": "jest --ci --runInBand --colors ./src/integration/pages/**/*.test.js",
-    "test:integration": "./scripts/killApp.sh && npm run build:local && (npm run start & wait-on -t 20000 http://localhost:7080/status) && jest --runInBand --colors ./src/integration/pages/**/*.test.js",
+    "test:integration": "./scripts/killApp.sh && npm run build:local && (npm run start & wait-on -t 20000 http://localhost:7080/status) && jest --no-cache --runInBand --colors ./src/integration/pages/**/*",
+    "test:integration:ci": "jest --no-cache --ci --runInBand --colors ./src/integration/pages/**/*",
```
Just checking, is `--no-cache` definitely needed? Just looked it up and it's a troubleshooting flag: https://jestjs.io/docs/en/troubleshooting#caching-issues . Just best to clean this sort of stuff up if it's not needed.

jroebu14 replied:
I changed the unit test snapshot serialiser so it was consistent with the integration tests. Didn't realise it was changing the snapshots like this. Reverted it back in https://github.com/bbc/simorgh/pull/6168/commits/3e9ead15136fe71de8ec84740c867af27508e2d9 and the issue is fixed. @andrewscfc

andrewscfc commented on src/app/containers/ATIAnalytics/canonical/__snapshots__/index.test.jsx.snap:
```diff
@@ -1,7 +1,53 @@
+    "withDomLvl1": true,
+    "xml": false,
+  },
+}
```
Why is this non-integration test snapshot changing?

**Same question for all below** 👇

jroebu14 replied:
Fixed

andrewscfc commented on src/integration/README.md:
```diff
@@ -127,23 +170,61 @@ You might then think it makes sense to only write integration tests but there ar
+## What is a snapshot?
 
-The Given-When-Then formula is a template intended to guide the writing of acceptance tests for a user story.
+This is a feature provided by Jest. It's not a snapshot of the graphical UI but a snapshot of a result that is typically returned from a function. This could even be HTML returned from DOM query selector e.g.
```
Just a change of emphasis suggestion, won't we be using `toMatchSnapshot` to match html most of the time?
```suggestion
This is a feature provided by Jest. It's not a snapshot of the graphical UI but a snapshot of a result that is typically returned from a function. This is often the HTML returned from DOM query selector e.g.
```

andrewscfc commented on src/integration/common/SEO.js:
```diff
@@ -1,151 +1,184 @@
+
+      expect(linkedDataEl).toBeInTheDocument();
+      expect(linkedDataEl.textContent).toBeTruthy();
+      expect(linkedDataEl.textContent).toMatchSnapshot();
```
this reads much better 🥇

andrewscfc commented on src/integration/common/core.canonical.js:
```diff
@@ -31,4 +31,13 @@ export default ({ service }) => {
+
+  //   expect(simorghDataScriptEl).toBeInTheDocument();
+  //   expect(simorghDataScriptEl).toMatchSnapshot();
+  // });
```
Looks like a valid test we should still have?

jroebu14 replied:
this test creates a huge snapshot for some files 😞

andrewscfc commented on src/integration/pages/articlePage/__snapshots__/mundo.test.js.snap:
```diff
@@ -1,2069 +0,0 @@
-// Jest Snapshot v1, https://goo.gl/fbAQLP
-
-exports[`Given I am on a Mundo Article page When I view the source code in the browser And using AMP I can see the server-rendered HTML 1`] = `
```
Glad to see the full page snapshots gone ‼️

Commit:
remove platform from global scope

Commit:
update readme

andrewscfc commented on src/integration/pages/articles/ampTests.js:
```diff
@@ -0,0 +1,6 @@
+
+export default () => {
+  runCoreAmpTests();
+  // runAmpAnalyticsTests();
```
comment

Commit:
Merge branch 'latest' into refactor-integration-tests

Commit:
readme update

karinathomasbbc commented on package.json:
```diff
@@ -33,8 +33,8 @@
     "test:unit:watch": "npm run test:unit -- --watch",
-    "test:integration:ci": "jest --ci --runInBand --colors ./src/integration/pages/**/*.test.js",
-    "test:integration": "./scripts/killApp.sh && npm run build:local && (npm run start & wait-on -t 20000 http://localhost:7080/status) && jest --runInBand --colors ./src/integration/pages/**/*.test.js",
+    "test:integration": "./scripts/killApp.sh && npm run build:local && (npm run start & wait-on -t 20000 http://localhost:7080/status) && jest --no-cache --runInBand --colors ./src/integration/pages/**/*",
```
Might be safer to remove `--no-cache` (apologies I know I requested to add this to the command initially). None of the other commands which use jest use this param, and the docs state that: _Note: the cache should only be disabled if you are experiencing caching related problems. On average, disabling the cache makes Jest at least two times slower_ - https://jestjs.io/docs/en/cli.html#--cache

karinathomasbbc commented on package.json:
```diff
@@ -33,8 +33,8 @@
-    "test:integration:ci": "jest --ci --runInBand --colors ./src/integration/pages/**/*.test.js",
-    "test:integration": "./scripts/killApp.sh && npm run build:local && (npm run start & wait-on -t 20000 http://localhost:7080/status) && jest --runInBand --colors ./src/integration/pages/**/*.test.js",
+    "test:integration": "./scripts/killApp.sh && npm run build:local && (npm run start & wait-on -t 20000 http://localhost:7080/status) && jest --no-cache --runInBand --colors ./src/integration/pages/**/*",
+    "test:integration:ci": "jest --no-cache --ci --runInBand --colors ./src/integration/pages/**/*",
```
I think we can remove `--no-cache` from here too

karinathomasbbc commented on src/app/containers/ATIAnalytics/canonical/__snapshots__/index.test.jsx.snap:
```diff
@@ -1,7 +1,53 @@
-<noscript>
-  &lt;img height="1px" width="1px" alt="" style="position:absolute" src="https://foobar.com?key=value&amp;key2=value2&amp;x8=[simorgh-nojs]"/&gt;
-</noscript>
+initialize {
```
Do we have any idea why this particular snapshot is so different to the baseline? (I know this is not part of your changes, but the unit tests fail without this). I checked this file in latest, and it looks like the before diff: https://github.com/bbc/simorgh/blob/latest/src/app/containers/ATIAnalytics/canonical/__snapshots__/index.test.jsx.snap

karinathomasbbc commented on src/app/containers/App/__snapshots__/App.test.jsx.snap:
```diff
@@ -1,28 +1,3 @@
-  </h1>
-</App>
-`;
+exports[`App should return rendered routes 1`] = `ReactWrapper {}`;
```
Another example of a weird snapshot - it has removed much of the detail :(

karinathomasbbc commented on src/app/containers/ArticleFigure/__snapshots__/index.test.jsx.snap:
```diff
@@ -1,115 +1,342 @@
-  .c4 {
-    grid-column: 1 / span 6;
-  }
+initialize {
```
Janky snapshot :(

jroebu14 replied:
Fixed

karinathomasbbc commented on src/app/containers/Image/__snapshots__/index.test.jsx.snap:
```diff
@@ -1,137 +1,365 @@
-  .c2 {
-    grid-template-columns: repeat(6,1fr);
-  }
+initialize {
```
Janky snapshot :(

karinathomasbbc commented on src/app/containers/ImageWithPlaceholder/__snapshots__/index.test.jsx.snap:
```diff
@@ -24,29 +24,133 @@ exports[`ImageWithPlaceholder should not provide non-js fallback 1`] = `
-  padding-bottom: 56.25%;
-  width: 100%;
-  background-image: url(data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiPz48c3ZnIHdpZHRoPSI0NzkiIGhlaWdodD0iMTM2IiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPjxnIGZpbGw9Im5vbmUiIGZpbGwtcnVsZT0iZXZlbm9kZCI+PGcgZmlsbD0iI0Q1RDBDRCI+PGc+PHBhdGggZD0iTTQ0OS41NTYgOTEuNTUzbC0yLjMxIDEuNDFjLTExLjE2NyA2LjgxOC0yMy4zMTMgMTAuNDc0LTM0LjM0NiAxMC40Ny0yMy42MS0uMDktMzkuMTYzLTE0LjA4My0zO ...
+initialize {
```
Janky snapshot :(

karinathomasbbc commented on src/integration/README.md:
```diff
@@ -14,38 +14,91 @@ This will build and run the application that the tests will run against.
 
-    expect(headlineEl).toBeInTheDocument();
+    expect(h1El).toBeInTheDocument();
+    expect(h1El).toBeTruthy();
```
Should this be `expect(h1El.textContent).toBeTruthy();` (just looking at other patterns across the tests, and we seem to check that the text, rather than the component itself is truthy

andrewscfc commented on src/integration/pages/articles/mundo/__snapshots__/canonical.test.js.snap:
```diff
@@ -0,0 +1,61 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`CANONICAL Articles A11y I can see the skip to content link 1`] = `"Ir al contenido"`;
```
Ah cool, this is a great approach, snapshotting the values

karinathomasbbc commented on src/integration/README.md:
```diff
@@ -119,31 +161,65 @@ You might then think it makes sense to only write integration tests but there ar
+  expect(h1El).toBeInTheDocument(); // check the headline element is in the document
+  expect(h1El.getAttribute('id')).toBe('content'); // check for the id attribute
+  expect(h1El.getAttribute('tabindex')).toBe('-1'); // check for the tabindex attribute
+  expect(h1El).toBeTruthy(); // check there is some text inside the element
```
Should this be `expect(h1El.textContent).toBeTruthy()` as per my previous comment?

karinathomasbbc commented on src/integration/common/SEO.js:
```diff
@@ -1,151 +1,184 @@
+      expect(headingEl.textContent).toMatchSnapshot();
+    });
+
+    it('Canonical link', () => {
+      const canonicalEl = document.querySelector('head link[rel="canonical"]');
+
+      expect(canonicalEl.getAttribute('href')).toMatchSnapshot();
+    });
+
+    it('Robots meta tag', () => {
+      const robotsEl = document.querySelector('head meta[name="robots"]');
+      const robotsContent = robotsEl.getAttribute('content');
+
+      expect(robotsContent).toMatchSnapshot();
+    });
+
+    it('Lang attribute', () => {
+      const htmlEl = document.querySelector('html');
+
+      expect(htmlEl.getAttribute('lang')).toMatchSnapshot();
+    });
+
+    it('Dir attribute', () => {
+      const htmlEl = document.querySelector('html');
+
+      expect(htmlEl.getAttribute('dir')).toMatchSnapshot();
+    });
```
Do any of these need a `truthy` check for the value of the attribute, like the metatdata content tests below?

Commit:
remove platform global from eslint

Commit:
update snapshots

andrewscfc commented on src/integration/utils/getPageTypeFromTestPath.js:
```diff
@@ -0,0 +1,6 @@
+module.exports = testPath => {
+  const [pageType] = testPath.match(/(?<=\/integration\/pages\/).+?(?=\/)/g) ||
+    testPath.match(/(?<=\/__GENERATED_TEST_FILES__\/.+?\/).+?(?=\/)/g) || [''];
```
Do we still have `__GENERATED_TEST_FILES__` in this PR?

jroebu14 replied:
Nope. I fixed this

Commit:
Merge branch 'refactor-integration-tests' of github.com:bbc/simorgh into refactor-integration-tests

karinathomasbbc commented on src/integration/common/analytics.amp.js:
```diff
@@ -1,9 +1,10 @@
+      expect(
+        document.querySelector('amp-analytics script[type="application/json"]')
+          .textContent,
+      ).toMatch('https://logws1363.ati-host.net?');
```
Can we have a `toMatchSnapshot()` here instead - it outputs the contents of the ATI script which will be useful for testing on demand radio in future.

karinathomasbbc commented on src/integration/common/analytics.canonical.js:
```diff
@@ -1,12 +1,14 @@
+      expect(noscriptImage.tagName).toEqual('IMG');
+      expect(noscriptImage.getAttribute('width')).toEqual('1px');
+      expect(noscriptImage.getAttribute('height')).toEqual('1px');
+      expect(noscriptImage.getAttribute('src')).toMatch(
```
As above, can we use `toMatchSnapshot()` instead?

Commit:
Merge branch 'latest' into refactor-integration-tests

karinathomasbbc commented on src/integration/common/core.canonical.js:
```diff
@@ -31,4 +31,13 @@ export default ({ service }) => {
     expect(serviceScripts.length).toBe(1);
   });
+
+  // it('Page data used for client side app hydration', () => {
+  //   const simorghDataScriptEl = Array.from(
+  //     document.querySelectorAll('script'),
+  //   ).find((script) => script.textContent.startsWith('window.SIMORGH_DATA'));
+
+  //   expect(simorghDataScriptEl).toBeInTheDocument();
+  //   expect(simorghDataScriptEl).toMatchSnapshot();
+  // });
```
Would this work if we uncommented it?

jroebu14 replied:
yeh but it creates huge ugly snapshots. I'm just going to remove it.

andrewscfc reviewed (commented):
Great work on this PR the basic idea of snapshoting the values between tests is great and maintains cross service coverage. How long does the suite take to run btw?

There are some issues in the comments, I'd highlight the unrelated snapshots being changed for containers, not really sure what is going on there.

One final question, I'm not expecting this, but has any progress been made on testing clientside logic?

karinathomasbbc commented on src/integration/common/mediaPlayer.js:
```diff
@@ -0,0 +1,13 @@
+        const mediaPlayerEl = document.querySelector('iframe, amp-iframe');
+
+        expect(mediaPlayerEl).toBeInTheDocument();
+        expect(mediaPlayerEl.getAttribute('src')).toBeTruthy();
```
I think this is supposed to be the `title` attribute, instead of `src` (the snapshot is `"https://polling.test.bbc.co.uk/ws/av-embeds/media/bbc_amharic_radio/liveradio/am/amp"` when it should be `Audio player`, `Media player` (or the equivalent translation).

karinathomasbbc commented on src/integration/pages/liveRadio/crossPlatformTests.js:
```diff
@@ -0,0 +1,28 @@
+    expect(summaryEl.textContent).toMatchSnapshot();
+  });
+
+  it('I can see the audio player embed', () => {
+    const audioPlayerIframe = document.querySelector('iframe, amp-iframe');
+
+    expect(audioPlayerIframe.getAttribute('src')).toMatchSnapshot();
+  });
```
Is it worth moving this test to  `src/integration/common/mediaPlayer.js` (but rename to media player embed)?

karinathomasbbc commented on src/integration/common/SEO.js:
```diff
@@ -1,151 +1,184 @@
+
+      expect(linkedDataEl).toBeInTheDocument();
+      expect(linkedDataEl.textContent).toBeTruthy();
+      expect(linkedDataEl.textContent).toMatchSnapshot();
```
Can we use JSON.parse for the text content, so that it looks nice in the snapshot e.g. 
```
expect(JSON.parse(linkedDataEl.textContent)).toMatchSnapshot();
```

jroebu14 replied:
Nice suggestion

Commit:
remove comments

Commit:
remove code not related to PR

Commit:
Merge branch 'latest' into refactor-integration-tests

Commit:
Merge branch 'refactor-integration-tests' of github.com:bbc/simorgh into refactor-integration-tests

Commit:
revert unit test serialiser. update snapshots

karinathomasbbc commented:
This looks ace, much easier to read, and loving all of the snapshots, which means it will be super easy to maintain in future. Thanks for all the hard work put into this 👍 🏆

Commit:
read title of media player rather than src

Commit:
nicer snapshot output

Commit:
remove commented test

Commit:
update snapshots

Commit:
parse JSON in canonical analytics test

Commit:
parse JSON in AMP analytics test

Commit:
remove --no-cache flag from test command

Commit:
update readme

jroebu14 commented:
@andrewscfc 

> How long does the suite take to run btw?

When the app has built and the server is running(10-20 seconds or something like that) the tests then take less than 7 seconds to run

<img width="252" alt="Screenshot 2020-04-21 at 10 38 20" src="https://user-images.githubusercontent.com/4798332/79850739-3dac1700-83bc-11ea-9077-35abd3c956bf.png">

---

> There are some issues in the comments, I'd highlight the unrelated snapshots being changed for containers, not really sure what is going on there.

I changed the unit test snapshot serialiser so it was consistent with the integration tests. Didn't realise it was changing the snapshots like this. Reverted it back the issue is fixed

---

> One final question, I'm not expecting this, but has any progress been made on testing clientside logic?

Not yet unfortunately!

Commit:
tidy up media player embed tests

Commit:
fix slightly botched refactor and tests

karinathomasbbc commented on src/integration/common/mediaPlayerEmbed.js:
```diff
@@ -0,0 +1,19 @@
+export default () => {
+  describe('Media player embed', () => {
+    describe('A11y', () => {
+      it('I can see the audio player embed', () => {
```
Can we just rename this to `I can see the audio player embed` embed as it is used for both live radio and MAPs

jroebu14 replied:
do you mean `I can see the media player embed`?

karinathomasbbc replied:
Oops, yes (copy paste error!). Thanks for that

karinathomasbbc commented on src/integration/common/mediaPlayerEmbed.js:
```diff
@@ -0,0 +1,19 @@
+export default () => {
+  describe('Media player embed', () => {
+    describe('A11y', () => {
```
I believe the a11y describe belongs with the `I can read the media player title` test

Commit:
Merge branch 'latest' into refactor-integration-tests

karinathomasbbc commented on src/integration/common/core.amp.js:
```diff
@@ -9,28 +9,23 @@ export default () => {
+  it('AMP Consent config', () => {
     expect(
-      amp.document.querySelector(
+      document.querySelector(
```
Could we json parse the output here, to have a more readable snapshot?

Commit:
fix describe context

karinathomasbbc commented on src/integration/common/core.amp.js:
```diff
@@ -9,28 +9,23 @@ export default () => {
-    ).toBe(
-      '{"ISOCountryGroups":{"eea":["at","ax","be","bg","bl","cy","cz","de","dk","ea","ee","es","fi","fr","gb","gf","gg","gi","gp","gr","hr","hu","ic","ie","im","is","it","je","li","lt","lu","lv","mf","mq","mt","nc","nl","no","pf","pl","pm","pt","re","ro","se","si","sj","sk","tf","va","wf","yt"]}}',
-    );
+      document.querySelector('body amp-geo > script[type="application/json"]')
```
Could we json parse the output here, to have a more readable snapshot?

Commit:
fix test title

Commit:
parse json to make snapshots more readable

Commit:
Merge branch 'latest' into refactor-integration-tests

andrewscfc reviewed (approved):
Looks great, one question about the test configs that could be fixed in a follow-up PR if I am correct, happy to merge

Commit:
Merge branch 'latest' into refactor-integration-tests

Commit:
update snapshots

Commit:
Merge branch 'latest' into refactor-integration-tests

Commit:
make test canonical only

simonsinclair commented on package.json:
```diff
@@ -33,8 +33,8 @@
     "test:unit:watch": "npm run test:unit -- --watch",
-    "test:integration:ci": "jest --ci --runInBand --colors ./src/integration/pages/**/*.test.js",
-    "test:integration": "./scripts/killApp.sh && npm run build:local && (npm run start & wait-on -t 20000 http://localhost:7080/status) && jest --runInBand --colors ./src/integration/pages/**/*.test.js",
+    "test:integration": "./scripts/killApp.sh && npm run build:local && (npm run start & wait-on -t 20000 http://localhost:7080/status) && jest --runInBand --colors ./src/integration/pages/**/*",
```
Because `(npm run start...)` is backgrounded, it isn't terminated when `test:integration` completes. This prevents later run start's/dev's, with the following error:
 
```
Error: bind EADDRINUSE null:7080
    at listenOnMasterHandle (net.js:1380:18)
    at rr (internal/cluster/child.js:132:12)
    at Worker.<anonymous> (internal/cluster/child.js:99:7)
    at process.onInternalMessage (internal/cluster/utils.js:47:8)
    at process.emit (events.js:322:22)
    at emit (internal/child_process.js:876:12)
    at processTicksAndRejections (internal/process/task_queues.js:85:21)
```

Could we utilise `killApp` or write a `runTestIntegration` script that manages the later part of this command?

simonsinclair reviewed (approved):
This is super work, @jroebu14 and I don't want to delay it, but I did spot something we might want to fix.

Edit: 2 seconds too late. 😄
